src/compiler/checker.ts

```diff
@@ -8965,10 +8966,13 @@ namespace ts {
             if (left.kind === SyntaxKind.SuperKeyword) {
                 return true;
             }
-            // A protected property is accessible in the declaring class and classes derived from it
+            // A protected property is accessible if the property is within the declaring class or classes derived from it
+            const typeClassDeclaration = <ClassLikeDeclaration>getClassLikeDeclarationOfSymbol(type.symbol);
             if (!enclosingClass || !hasBaseType(enclosingClass, declaringClass)) {
-                error(node, Diagnostics.Property_0_is_protected_and_only_accessible_within_class_1_and_its_subclasses, symbolToString(prop), typeToString(declaringClass));
-                return false;
+                if (!isNodeWithinClass(node, typeClassDeclaration)) {

```

Comment:
i think this should be a loop walking up and repeate the check for each enclosing class.

I would consider adding a function like `forEachEnclosingClass` and wrapp the exisitng consitions to the call to of that function:

e.g.: `if (!enclosingClass || ! forEachEnclosingClass(node, enclosingClass=> hasBaseType(enclosingClass, declaringClass))) {`

where `forEachEnclosingClass` is defined as:

```ts
    function forEachEnclosingClass<T>(node:Node, callback: (node:Node) => T):T {
        let result: T;

        while (true) {
            node = getContainingClass(node);
            if (!node) break;
            if (result = callback(node)) break;
        }

        return result;
    }
```
and `isNodeWithinClass` can be defined as:

```ts
    function isNodeWithinClass(node: Node, classDeclaration: ClassLikeDeclaration) {
        return !!forEachEnclosingClass(node, n => n === classDeclaration);
    }
```